frame/evm/precompile/contracts/encoder/src/lib.rs

```diff
@@ -47,30 +57,82 @@ where
 		}
-		let selector = &input[0..SELECTOR];
-		let call: T::Call = match selector {
-			_ if selector == <from_substrate_issuing::Pallet<T>>::digest() => {
-				from_substrate_issuing::Call::<T>::asset_burn_event_handle(input.to_vec()).into()
+		let pallet_digest = &input[0..PALLET_DIG_LEN];
+		let method_digest = &input[PALLET_DIG_LEN..ACTION_LEN];
+		let output = match pallet_digest {
+			_ if pallet_digest == <from_substrate_issuing::Pallet<T>>::digest() => {
+				match method_digest {
+					_ if Self::match_digest(method_digest, BURN_AND_REMOTE_UNLOCK_METHOD) => {
+						let call: T::Call =
+							from_substrate_issuing::Call::<T>::asset_burn_event_handle(
+								input[ACTION_LEN..].to_vec(),
+							)
+							.into();
+						call.encode()
+					}
+					// this method comes from mapping-token-factory, we ignore it by a empty method
+					_ if Self::match_digest(method_digest, TOKEN_REGISTER_RESPONSE_METHOD) => {
+						Vec::new()
+					}
+					_ if Self::match_digest(method_digest, READ_LATEST_MESSAGE_ID_METHOD) => {
+						<T as from_substrate_issuing::Config>::MessageSender::latest_message_id()
+							.to_vec()
+					}
+					_ => {
+						return Err(ExitError::Other(
+							"No such method in pallet substrate issuing".into(),
+						));
+					}
+				}
 			}
-			_ if selector == <from_ethereum_issuing::Pallet<T>>::digest() => {
-				from_ethereum_issuing::Call::<T>::mapping_factory_event_handle(input.to_vec())
-					.into()
+			_ if pallet_digest == <from_ethereum_issuing::Pallet<T>>::digest() => {
+				match method_digest {
+					_ if Self::match_digest(method_digest, TOKEN_REGISTER_RESPONSE_METHOD) => {
+						let call: T::Call =
+							from_ethereum_issuing::Call::<T>::register_response_from_contract(
+								input[ACTION_LEN..].to_vec(),
+							)
+							.into();
+						call.encode()
+					}
+					_ if Self::match_digest(method_digest, BURN_AND_REMOTE_UNLOCK_METHOD) => {
+						let call: T::Call =
+							from_ethereum_issuing::Call::<T>::deposit_burn_token_event_from_precompile(
+								input[ACTION_LEN..].to_vec(),
+							)
+							.into();
+						call.encode()
+					}
+					// todo, when ethereum support message confirm, we need give this message id
+					_ if Self::match_digest(method_digest, READ_LATEST_MESSAGE_ID_METHOD) => {
+						Vec::new()
+					}
+					_ => {
+						return Err(ExitError::Other(
+							"No such method in pallet ethereum issuing".into(),
+						));
+					}
+				}
 			}
 			_ => {
-				return Err(ExitError::Other(
-					"No wrapper method at selector given selector".into(),
-				));
+				return Err(ExitError::Other("No valid pallet digest found".into()));
 			}
 		};
-		// TODO: The cost should not be zero
+		// estimate a cost for this encoder process
 		Ok(PrecompileOutput {
 			exit_status: ExitSucceed::Stopped,
-			cost: 0,
-			output: call.encode(),
+			cost: 20000,
+			output,
 			logs: Default::default(),
 		})
 	}
 }
+
+impl<T> DispatchCallEncoder<T> {
```

Comment:
Actually, we do not need this `impl`, `match_digest` has no relationship with `DispatchCallEncoder`, right?

Reply:
Yes, a normal function also works.